synapse/handlers/oidc_handler.py

```diff
@@ -280,7 +290,7 @@ async def load_jwks(self, force: bool = False) -> JWKS:
                     ]
                 }
         """
-        if self._uses_userinfo:
+        if token is not None and self._uses_userinfo(token):
```

Comment:
it looks to me like this condition will always be False? (can you tell me when it will be True?)

apart from it making the code unclear, this is problematic functionally: it means we will *always* attempt to fetch the `jwks_uri` during startup, even if we will not be using jwt signing. I'm no expert, but surely providers are not obliged to provide a `jwks_uri`?

Reply:
You are right: This will always be False during startup. I don't think we can do any better for `auto` because we cannot know whether the provider will return an `id_token`.

Potential solutions:
- Catch any errors when prefetching jwks during startup (ugly!).
- Don't prefetch.
- Don't support `auto` (or only use info we can know in advance).

Reply:
>  This will always be False during startup. 

not just during startup: *always*.

Reply:
If `load_jwks` is called with a token that doesn't have any `id_token`, this should be True, right? It will be False if we don't pass any token to `load_jwks` but that should only happen during startup.

This is tested by `test_load_jwks` (after comment `Return empty key set if JWKS are not used`).